src/main/java/com/fasterxml/jackson/core/JsonGenerator.java

```diff
@@ -2636,13 +2638,15 @@ protected void _copyCurrentContents(JsonParser p) throws IOException
             }
             case ID_NUMBER_FLOAT:
             {
-                NumberType n = p.getNumberType();
-                if (n == NumberType.BIG_DECIMAL) {
-                    writeNumber(p.getDecimalValue());
-                } else if (n == NumberType.FLOAT) {
-                    writeNumber(p.getFloatValue());
+                Number n = p.getNumberValueExact();
+                if (n instanceof Float) {
+                    writeNumber((Float) n);
+                } else if (n instanceof BigDecimal) {
+                    final BigDecimal bd = (BigDecimal) n;
+                    p.streamReadConstraints().validateBigIntegerScale(bd.scale());
+                    writeNumber(bd);
                 } else {
-                    writeNumber(p.getDoubleValue());
+                    writeNumber(n.longValue());
```

Comment:
Also shouldn't this be "n.doubleValue()"...

Reply:
fixed